lib/index.js

```diff
@@ -17,183 +17,174 @@ internals.defaults = {
-        return Promise.resolve();
-    }
+    async start() {
 
-    // Return a promise that is resolved when everything is ready
-    return new Promise((resolve, reject) => {
+        if (this.settings.client) {
+            this.client = this.settings.client;
+        }
 
-        let client;
+        if (this.client) {
+            return;
+        }
 
-        const options = {
-            password: this.settings.password,
-            db: this.settings.database || this.settings.db,
-            tls: this.settings.tls
-        };
+        // Return a promise that is resolved when everything is ready
+        return await new Promise((resolve, reject) => {
 
-        if (this.settings.sentinels && this.settings.sentinels.length) {
-            options.sentinels = this.settings.sentinels;
-            options.name = this.settings.sentinelName;
-            client = Redis.createClient(options);
-        }
-        else if (this.settings.url) {
-            client = Redis.createClient(this.settings.url, options);
-        }
-        else if (this.settings.socket) {
-            client = Redis.createClient(this.settings.socket, options);
-        }
-        else {
-            client = Redis.createClient(this.settings.port, this.settings.host, options);
-        }
+            let client;
 
-        // Listen to errors
-        client.on('error', (err) => {
+            const options = {
+                password: this.settings.password,
+                db: this.settings.database || this.settings.db,
+                tls: this.settings.tls
+            };
 
-            if (!this.client) {             // Failed to connect
-                client.end(false);
-                return reject(err);
+            if (this.settings.sentinels && this.settings.sentinels.length) {
+                options.sentinels = this.settings.sentinels;
+                options.name = this.settings.sentinelName;
+                client = Redis.createClient(options);
+            }
+            else if (this.settings.url) {
+                client = Redis.createClient(this.settings.url, options);
+            }
+            else if (this.settings.socket) {
+                client = Redis.createClient(this.settings.socket, options);
+            }
+            else {
+                client = Redis.createClient(this.settings.port, this.settings.host, options);
             }
-        });
-
-        // Wait for connection
-        client.once('ready', () => {
 
-            this.client = client;
-            return resolve();
-        });
-    });
-};
+            // Listen to errors
+            client.on('error', (err) => {
 
+                if (!this.client) {             // Failed to connect
+                    client.end(false);
+                    return reject(err);
+                }
+            });
 
-internals.Connection.prototype.stop = async function () {
+            // Wait for connection
+            client.once('ready', () => {
```

Comment:
What do you think about using `await ioredis.connect()` instead of listening for the event? This is new in ioredis v4 and resolves to the `ready` status ( resolved to `connect` in ioredis v3)

Reply:
This doesn't seem to work correctly, it causes a lot of unhandled error events and causes tests to fail.

Test code:
```js
        const options = {
            password: this.settings.password,
            db: this.settings.database || this.settings.db,
            tls: this.settings.tls,
            lazyConnect: true, // required so we can call `connect`
        };

        if (this.settings.sentinels && this.settings.sentinels.length) {
            options.sentinels = this.settings.sentinels;
            options.name = this.settings.sentinelName;
            client = new Redis(options);
        }
        else if (this.settings.url) {
            client = new Redis(this.settings.url, options);
        }
        else if (this.settings.socket) {
            client = new Redis(this.settings.socket, options);
        }
        else {
            client = new Redis(this.settings.port, this.settings.host, options);
        }

        await client.connect();
        this.client = client;
```

Reply:
I have to assume `.connect` doesn't perform proper error handling, which makes the method mostly useless unfortunately.